File: Python/0201-0250/238. Product of Array Except Self.py
# https://leetcode.com/problems/product-of-array-except-self/description/

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        size = len(nums)
        # calculate the left and the right sum
        left = []
        right = []

        # Calculate the left sum
        left.append(1)
        for index in range(1, size):
            left.append(left[index-1] * nums[index-1])

        # calculate the right sum
        right.append(1)
        for index in reversed(range(0, size-1)):
            right.insert(0, right[0] * nums[index+1])

        # calculate the result 
        result = []
        for index in range(0, size):
            result.append(left[index] * right[index])

        return result;

File: Java/0201-0250/238. Product of Array Except Self.java
// https://leetcode.com/problems/product-of-array-except-self/description/

class Solution {
    public int[] productExceptSelf(int[] nums) {
        int size = nums.length;
        // calculate the left and the right sum
        int[] left = new int[size];
        int[] right = new int[size];

        // Calculate the left sum
        left[0] = 1;
        for(int index=1; index<size; index++)
            left[index] = left[index-1] * nums[index-1];

        // calculate the right sum
        right[size-1] = 1;
        for(int index = size-2; index>=0; index--)
            right[index] = right[index+1] * nums[index+1];

        // calculate the result 
        int[] result = new int[size];
        for(int index=0; index<size; index++)
            result[index] = left[index] * right[index];

        return result;
    }
}
